# Eat Safe, Love

## Part 1: Database and Jupyter Notebook Set Up

Import the data provided in the `establishments.json` file from your Terminal. Name the database `uk_food` and the collection `establishments`.

Within this markdown cell, copy the line of text you used to import the data from your Terminal. This way, future analysts will be able to repeat your process.

e.g.: Import the dataset with `YOUR IMPORT TEXT HERE`

2024-11-08T02:42:48.978-0600    connected to: mongodb://localhost/
2024-11-08T02:42:50.114-0600    39779 document(s) imported successfully. 0 document(s) failed to import.

## Database Verification
- Verified databases with `show dbs`:
  - `uk_food` is present.
- Switched to `uk_food` with `use uk_food`.
- Verified collection with `show collections`:
  - `establishments` is present.
- Verified document count with `db.establishments.countDocuments()`:
  - Total documents: 39,779.
- Displayed a sample document with `db.establishments.findOne()`:

In [18]:
# Import dependencies
from pymongo import MongoClient
from pprint import pprint
from pymongo import UpdateOne

In [2]:
# Create an instance of MongoClient
mongo = MongoClient(port=27017)

In [ ]:
# confirm that our new database was created
# List all databases
show dbs
# Output:
admin       40.00 KiB
config      72.00 KiB
local       72.00 KiB
myDatabase  40.00 KiB
uk_food      8.37 MiB

# Switch to the `uk_food` database
use uk_food

# List collections in the `uk_food` database
show collections
# Output:
establishments

# Count the number of documents in the `establishments` collection
db.establishments.countDocuments()
# Output:
39779

# Display a sample document
db.establishments.findOne()
# Output:
{
  "_id": ObjectId('...'),
  "BusinessName": "Refreshment Kiosk",
  "BusinessType": "Restaurant/Cafe/Canteen",
  ...
}

In [3]:
# assign the uk_food database to a variable name
db = mongo['uk_food']

In [4]:
# Verify that the database and collection exist
print("Databases:", mongo.list_database_names())
print("Collections in 'uk_food':", db.list_collection_names())


Databases: ['admin', 'config', 'local', 'myDatabase', 'uk_food']
Collections in 'uk_food': ['establishments']


In [5]:
# Preview one document in the establishments collection
establishments = db['establishments']
pprint(establishments.find_one())

{'AddressLine1': 'The Bay',
 'AddressLine2': 'St Margarets Bay',
 'AddressLine3': 'Kent',
 'AddressLine4': '',
 'BusinessName': 'Refreshment Kiosk',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'ChangesByServerID': 0,
 'Distance': 4587.347174863443,
 'FHRSID': 254719,
 'LocalAuthorityBusinessID': 'PI/000069980',
 'LocalAuthorityCode': '182',
 'LocalAuthorityEmailAddress': 'publicprotection@dover.gov.uk',
 'LocalAuthorityName': 'Dover',
 'LocalAuthorityWebSite': 'http://www.dover.gov.uk/',
 'NewRatingPending': False,
 'Phone': '',
 'PostCode': 'CT15 6DY',
 'RatingDate': '2022-03-24T00:00:00',
 'RatingKey': 'fhrs_5_en-gb',
 'RatingValue': '5',
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('672dcf0860c906defe5b86f1'),
 'geocode': {'latitude': '51.152225', 'longitude': '1.387974'},
 'links': [{'href': 'https://api.ratings.food.gov.uk/establishments/254719',
            'rel': 'self'}],
 'meta': {'dataSource': None,
          'extractDate': '0001-01-01T0

In [ ]:
# review the collections in our new database


In [ ]:
# review the collections in our new database


#### Reviewing Collections in the Database

To list the collections in the `uk_food` database, the following command was run in the MongoDB shell:

```javascript
show collections

Output:

establishments



In [ ]:
# review a document in the establishments collection


#### Reviewing a Document in the `establishments` Collection

To preview a document in the `establishments` collection, the following command was run in the MongoDB shell:

```javascript
db.establishments.findOne()

{
  "_id": ObjectId('672dcf0860c906defe5b86f1'),
  "FHRSID": 254719,
  "ChangesByServerID": 0,
  "LocalAuthorityBusinessID": "PI/000069980",
  "BusinessName": "Refreshment Kiosk",
  "BusinessType": "Restaurant/Cafe/Canteen",
  ...
}

In [ ]:
# assign the collection to a variable
establishments = db['establishments']

## Part 2: Update the Database

1. An exciting new halal restaurant just opened in Greenwich, but hasn't been rated yet. The magazine has asked you to include it in your analysis. Add the following restaurant "Penang Flavours" to the database.

In [6]:
# Create a dictionary for the new restaurant data
new_restaurant = {
    "BusinessName": "Penang Flavours",
    "BusinessType": "Restaurant/Cafe/Canteen",
    "AddressLine1": "123 Greenwich High Road",
    "AddressLine2": "",
    "AddressLine3": "",
    "AddressLine4": "",
    "PostCode": "SE10 8JL",
    "LocalAuthorityName": "Greenwich",
    "RatingValue": "Awaiting Inspection",
    "latitude": "51.4827",
    "longitude": "-0.0077"
}
establishments.insert_one(new_restaurant)
pprint(establishments.find_one({"BusinessName": "Penang Flavours"}))

{'AddressLine1': '123 Greenwich High Road',
 'AddressLine2': '',
 'AddressLine3': '',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'LocalAuthorityName': 'Greenwich',
 'PostCode': 'SE10 8JL',
 'RatingValue': 'Awaiting Inspection',
 '_id': ObjectId('672de5062a4468fd2f3eb37f'),
 'latitude': '51.4827',
 'longitude': '-0.0077'}


In [7]:
# Create a dictionary for the new restaurant
new_restaurant = {
    "BusinessName": "Penang Flavours",
    "BusinessType": "Restaurant/Cafe/Canteen",
    "AddressLine1": "123 Greenwich High Road",
    "AddressLine2": "",
    "AddressLine3": "",
    "AddressLine4": "",
    "PostCode": "SE10 8JL",
    "LocalAuthorityName": "Greenwich",
    "RatingValue": "Awaiting Inspection",
    "latitude": "51.4827",
    "longitude": "-0.0077"
}


In [8]:
# Insert the new restaurant into the collection
establishments.insert_one(new_restaurant)

InsertOneResult(ObjectId('672de5d52a4468fd2f3eb380'), acknowledged=True)

In [9]:
# Check that the new restaurant was inserted
pprint(establishments.find_one({"BusinessName": "Penang Flavours"}))

{'AddressLine1': '123 Greenwich High Road',
 'AddressLine2': '',
 'AddressLine3': '',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'LocalAuthorityName': 'Greenwich',
 'PostCode': 'SE10 8JL',
 'RatingValue': 'Awaiting Inspection',
 '_id': ObjectId('672de5062a4468fd2f3eb37f'),
 'latitude': '51.4827',
 'longitude': '-0.0077'}


2. Find the BusinessTypeID for "Restaurant/Cafe/Canteen" and return only the `BusinessTypeID` and `BusinessType` fields.

In [10]:
# Find the BusinessTypeID for "Restaurant/Cafe/Canteen" and return only the BusinessTypeID and BusinessType fields
result = establishments.find_one({"BusinessType": "Restaurant/Cafe/Canteen"}, {"BusinessTypeID": 1, "BusinessType": 1})
pprint(result)
business_type_id = result["BusinessTypeID"]

{'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 '_id': ObjectId('672dcf0860c906defe5b86f1')}


3. Update the new restaurant with the `BusinessTypeID` you found.

In [11]:
# Update the new restaurant with the correct BusinessTypeID
establishments.update_one(
    {"BusinessName": "Penang Flavours"},
    {"$set": {"BusinessTypeID": business_type_id}}
)
pprint(establishments.find_one({"BusinessName": "Penang Flavours"}))

{'AddressLine1': '123 Greenwich High Road',
 'AddressLine2': '',
 'AddressLine3': '',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'LocalAuthorityName': 'Greenwich',
 'PostCode': 'SE10 8JL',
 'RatingValue': 'Awaiting Inspection',
 '_id': ObjectId('672de5062a4468fd2f3eb37f'),
 'latitude': '51.4827',
 'longitude': '-0.0077'}


In [12]:
# Confirm that the new restaurant was updated
updated_restaurant = establishments.find_one({"BusinessName": "Penang Flavours"})
pprint(updated_restaurant)

{'AddressLine1': '123 Greenwich High Road',
 'AddressLine2': '',
 'AddressLine3': '',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'LocalAuthorityName': 'Greenwich',
 'PostCode': 'SE10 8JL',
 'RatingValue': 'Awaiting Inspection',
 '_id': ObjectId('672de5062a4468fd2f3eb37f'),
 'latitude': '51.4827',
 'longitude': '-0.0077'}


4. The magazine is not interested in any establishments in Dover, so check how many documents contain the Dover Local Authority. Then, remove any establishments within the Dover Local Authority from the database, and check the number of documents to ensure they were deleted.

In [13]:
# Find how many documents have LocalAuthorityName as "Dover"
dover_count = establishments.count_documents({"LocalAuthorityName": "Dover"})
print("Documents with Dover Local Authority:", dover_count)


Documents with Dover Local Authority: 994


In [14]:
# Delete all documents where LocalAuthorityName is "Dover"
establishments.delete_many({"LocalAuthorityName": "Dover"})

DeleteResult({'n': 994, 'ok': 1.0}, acknowledged=True)

In [15]:
# Check how many documents have LocalAuthorityName as "Dover"
dover_count = establishments.count_documents({"LocalAuthorityName": "Dover"})
print(f"Number of documents with Dover as LocalAuthorityName: {dover_count}")

Number of documents with Dover as LocalAuthorityName: 0


In [16]:
# Check if any remaining documents include Dover
dover_document = establishments.find_one({"LocalAuthorityName": "Dover"})
pprint(dover_document)

None


In [17]:
# Check that other documents remain with 'find_one'
remaining_document = establishments.find_one()
pprint(remaining_document)

{'AddressLine1': 'East Cliff Pavilion',
 'AddressLine2': 'Wear Bay Road',
 'AddressLine3': 'Folkestone',
 'AddressLine4': 'Kent',
 'BusinessName': 'The Pavilion',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'ChangesByServerID': 0,
 'Distance': 4591.765489457773,
 'FHRSID': 1043695,
 'LocalAuthorityBusinessID': 'PI/000073616',
 'LocalAuthorityCode': '188',
 'LocalAuthorityEmailAddress': 'foodteam@folkestone-hythe.gov.uk',
 'LocalAuthorityName': 'Folkestone and Hythe',
 'LocalAuthorityWebSite': 'http://www.folkestone-hythe.gov.uk',
 'NewRatingPending': False,
 'Phone': '',
 'PostCode': 'CT19 6BL',
 'RatingDate': '2018-04-04T00:00:00',
 'RatingKey': 'fhrs_5_en-gb',
 'RatingValue': '5',
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('672dcf0860c906defe5b89d0'),
 'geocode': {'latitude': '51.083812', 'longitude': '1.195625'},
 'links': [{'href': 'https://api.ratings.food.gov.uk/establishments/1043695',
            'rel': 'self'}],
 'meta': {'dataSource': 

5. Some of the number values are stored as strings, when they should be stored as numbers.

Use `update_many` to convert `latitude` and `longitude` to decimal numbers.

In [19]:
# Change the data type from String to Decimal for longitude and latitude
establishments.bulk_write([
    UpdateOne({"latitude": {"$type": "string"}}, [{"$set": {"latitude": {"$toDecimal": "$latitude"}}}]),
    UpdateOne({"longitude": {"$type": "string"}}, [{"$set": {"longitude": {"$toDecimal": "$longitude"}}}])
])

BulkWriteResult({'writeErrors': [], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 0, 'nMatched': 2, 'nModified': 2, 'nRemoved': 0, 'upserted': []}, acknowledged=True)

Use `update_many` to convert `RatingValue` to integer numbers.

In [20]:
# List of non-1-to-5 rating values
non_ratings = ["AwaitingInspection", "Awaiting Inspection", "AwaitingPublication", "Pass", "Exempt"]

# Set non-1-to-5 RatingValues to null
result = establishments.update_many(
    {"RatingValue": {"$in": non_ratings}},  # Filter for invalid rating values
    {"$set": {"RatingValue": None}}         # Update action
)

# Output the number of updated documents
print(f"Documents updated: {result.modified_count}")

Documents updated: 4093


In [21]:
# Verify that invalid RatingValues were set to null
updated_document = establishments.find_one({"RatingValue": None})
pprint(updated_document)

{'AddressLine1': 'Shepway Sports Trust',
 'AddressLine2': 'The Stade',
 'AddressLine3': 'Folkestone',
 'AddressLine4': 'Kent',
 'BusinessName': 'Folkestone Seasports',
 'BusinessType': 'Retailers - other',
 'BusinessTypeID': 4613,
 'ChangesByServerID': 0,
 'Distance': 4591.923505280542,
 'FHRSID': 1285095,
 'LocalAuthorityBusinessID': 'PI/000077988',
 'LocalAuthorityCode': '188',
 'LocalAuthorityEmailAddress': 'foodteam@folkestone-hythe.gov.uk',
 'LocalAuthorityName': 'Folkestone and Hythe',
 'LocalAuthorityWebSite': 'http://www.folkestone-hythe.gov.uk',
 'NewRatingPending': False,
 'Phone': '',
 'PostCode': '',
 'RatingDate': '1901-01-01T00:00:00',
 'RatingKey': 'fhrs_awaitinginspection_en-gb',
 'RatingValue': None,
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('672dcf0860c906defe5b89ff'),
 'geocode': {'latitude': '51.0805695006979', 'longitude': '1.18794649942731'},
 'links': [{'href': 'https://api.ratings.food.gov.uk/establishments/1285095',
            'rel': 'self'}

In [22]:
# Change the data type from String to Integer for RatingValue
non_ratings = ["AwaitingInspection", "Awaiting Inspection", "AwaitingPublication", "Pass", "Exempt"]
establishments.update_many({"RatingValue": {"$in": non_ratings}}, [{"$set": {"RatingValue": None}}])

establishments.update_many({"RatingValue": {"$type": "string"}}, [{"$set": {"RatingValue": {"$toInt": "$RatingValue"}}}])

UpdateResult({'n': 34694, 'nModified': 34694, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)

In [26]:
# Find one document and check the data types of latitude and longitude
document = establishments.find_one({}, {"geocode.latitude": 1, "geocode.longitude": 1})
latitude = document["geocode"]["latitude"]
longitude = document["geocode"]["longitude"]

print(f"Latitude: {latitude}, Type: {type(latitude)}")
print(f"Longitude: {longitude}, Type: {type(longitude)}")

Latitude: 51.083812, Type: <class 'str'>
Longitude: 1.195625, Type: <class 'str'>


In [27]:
# Find one document and check the data type of RatingValue
document = establishments.find_one({}, {"RatingValue": 1})
rating_value = document["RatingValue"]

print(f"RatingValue: {rating_value}, Type: {type(rating_value)}")

RatingValue: 5, Type: <class 'int'>


In [28]:
# Check that the coordinates and rating value are now numbers
# Check the data type of latitude and longitude for one document
document = establishments.find_one({"latitude": {"$exists": True}, "longitude": {"$exists": True}})
print(f"Latitude: {document['latitude']} (Type: {type(document['latitude'])})")
print(f"Longitude: {document['longitude']} (Type: {type(document['longitude'])})")

Latitude: 51.4827 (Type: <class 'bson.decimal128.Decimal128'>)
Longitude: -0.0077 (Type: <class 'bson.decimal128.Decimal128'>)


In [29]:
# Check the data type of RatingValue for one document
rating_document = establishments.find_one({"RatingValue": {"$exists": True}})
print(f"RatingValue: {rating_document['RatingValue']} (Type: {type(rating_document['RatingValue'])})")

RatingValue: 5 (Type: <class 'int'>)
